File: src/II079.cpp
#include <iostream>
#include <vector>
using namespace std;

// 说好了搜索与回溯，题解怎么是迭代
class Solution {
public:
  vector<vector<int>> subsets(vector<int>& nums) {
    vector<int> cur;
    vector<vector<int>> res;
    int n = nums.size();
    for (int mask = 0; mask < 1 << n; ++mask) {
      cur.clear();
      for (int i = 0; i < n; ++i) {
        if (mask & (1 << i)) {
          cur.push_back(nums[i]);
        }
      }
      res.push_back(cur);
    }
    return res;
  }
};

File: src/II079.rs
struct Solution;

impl Solution {
    pub fn subsets(nums: Vec<i32>) -> Vec<Vec<i32>> {
        fn get_res(nums: &Vec<i32>, res: &mut Vec<Vec<i32>>, cur: &mut Vec<i32>, start_i: usize) {
            res.push(cur.clone());
            for i in start_i..nums.len() {
                cur.push(nums[i]);
                get_res(nums, res, cur, i + 1);
                cur.pop();
            }
        }
        let mut res = vec![];
        get_res(&nums, &mut res, &mut vec![], 0);
        res
    }
}
